ddl/placement_rule_test.go

```diff
@@ -194,13 +198,107 @@ func (s *testPlacementSuite) TestPlacementBuild(c *C) {
 				},
 			},
 		},
+
+		{
+			input: []*ast.PlacementSpec{
+				{
+					Role:        ast.PlacementRoleVoter,
+					Tp:          ast.PlacementAdd,
+					Replicas:    3,
+					Constraints: `["+  zone=sh", "-zone = bj"]`,
+				},
+				{
+					Role: ast.PlacementRoleVoter,
+					Tp:   ast.PlacementDrop,
+				},
+			},
+			output: []*placement.RuleOp{
+				{
+					Action:           placement.RuleOpDel,
+					DeleteByIDPrefix: true,
+					Rule: &placement.Rule{
+						GroupID: placement.RuleDefaultGroupID,
+						Role:    placement.Voter,
+					},
+				},
+			},
+		},
+
+		{
+			input: []*ast.PlacementSpec{
+				{
+					Role: ast.PlacementRoleLearner,
+					Tp:   ast.PlacementDrop,
+				},
+				{
+					Role: ast.PlacementRoleVoter,
+					Tp:   ast.PlacementDrop,
+				},
+			},
+			output: []*placement.RuleOp{
+				{
+					Action:           placement.RuleOpDel,
+					DeleteByIDPrefix: true,
+					Rule: &placement.Rule{
+						GroupID: placement.RuleDefaultGroupID,
+						Role:    placement.Voter,
+					},
+				},
+				{
+					Action:           placement.RuleOpDel,
+					DeleteByIDPrefix: true,
+					Rule: &placement.Rule{
+						GroupID: placement.RuleDefaultGroupID,
+						Role:    placement.Learner,
+					},
+				},
+			},
+		},
+
+		{
+			input: []*ast.PlacementSpec{
+				{
+					Role:        ast.PlacementRoleLearner,
+					Tp:          ast.PlacementAdd,
+					Replicas:    3,
+					Constraints: `["+  zone=sh", "-zone = bj"]`,
+				},
+				{
+					Role: ast.PlacementRoleVoter,
+					Tp:   ast.PlacementDrop,
+				},
+			},
+			output: []*placement.RuleOp{
+				{
+					Action:           placement.RuleOpDel,
+					DeleteByIDPrefix: true,
+					Rule: &placement.Rule{
+						GroupID: placement.RuleDefaultGroupID,
+						Role:    placement.Voter,
+					},
+				},
+				{
+					Action: placement.RuleOpAdd,
+					Rule: &placement.Rule{
+						GroupID:  placement.RuleDefaultGroupID,
+						Role:     placement.Learner,
+						Override: true,
+						Count:    3,
+						LabelConstraints: []placement.LabelConstraint{
+							{Key: "zone", Op: "in", Values: []string{"sh"}},
+							{Key: "zone", Op: "notIn", Values: []string{"bj"}},
+						},
+					},
+				},
+			},
+		},
 	}
 	for k, t := range tests {
 		out, err := buildPlacementSpecs(t.input)
-		c.Logf("test %d\n", k)
+		c.Logf("test %d", k)
```

Comment:
Some of the logs can be removed. Only log it when the result is not expected.

Reply:
it is moved into the fail block.